server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java

```diff
@@ -369,7 +392,7 @@ private void addSegments(Collection<DataSegment> segments, final DataSegmentChan
                     numSegments,
                     segment.getId()
                 );
-                loadSegment(segment, callback, config.isLazyLoadOnStart());
+                loadSegment(segment, callback, config.isLazyLoadOnStart(), loadSegmentsIntoPageCacheOnBootstrapExec);
```

Comment:
Hi Parag, thanks for the comment! To clarify, the point of the PR is to provide best effort precache of segments but not 100% guaranteed precache of segments. The reason is we don't want to sacrifice availability.

Image we have a data source with one replica configured and the host serving it dies, then we want the missing segments to be available ASAP on other replacement host. This is the case of loading segments into a complete new host with its page cache empty of the segments to load. Copying segments to null stream can take some time and if you have a lot segments, it can easily take more than 10 mins to complete the full read into page cache which is too long period of unavailability. In production, we tried the synchronous loading before announcing the segment, and it was indeed too slow. Another case is if we just restart the Druid historical process on the same host, then the process of reading segments into null stream is pretty fast as the segments are already cached in page cache by OS but it will still take some time compared with announcing segments directly after download.

That said, the strategy of announce segment immediately after download and asynchronously read into null stream afterwards still have value: without the change, since OS will only mmap the portion of segments a query needs, even after days of serving a segment in a historical, a different query hitting a different portion of segment will still need to trigger disk reads. With the change, we can ensure after the first 10 mins or so after downloading the segments into a historical (depending on the number of segments), all subsequent queries won't hit disk.

Reply:
I understand your point I am just worried that when a historical is warming up the query performance at that time will be worse than performance after usual boot up as its already doing disk reads for warming up and doing more disk reads for serving query. 

Can we introduce a flag to enable synchronous warm up which when enabled historical only announces itself after everything is read once. When warm up is enabled default behaviour will be async as in this PR but can be changed to sync warm up when boot up time is not a concern. How does this sound to you ? Thanks

Reply:
Sg, thanks!

Reply:
@wjhypo any updates on when can you make the changes ?

Reply:
@pjain1 Hi Parag, I hope I can get something by end of next week

Reply:
Thanks

Reply:
gentle reminder

Reply:
@pjain1 Hi Parag, sorry about the delay. Sorry I still didn't get time to make the change and verify the sync approach in a test cluster. Not sure if any user is in need of the sync approach. We only verified the async approach in production. At this point, it would be better if someone in need of the sync approach can help make the change and dedicate some time to verify the change. It currently doesn't align with our priority at work so I won't be able to make the change this year.

Reply:
@wjhypo no issues, I think we can get it in as in current condition. Can work on the sync behaviour later.

Few questions - Do you run this on your prod env ? Hows the performance after starting of the historical process and before warm up is complete as process will be doing double reads for warming up as well as serving query ? Do you see more of timeouts during this time ? Is it even usable at that time ? Thanks

Reply:
@wjhypo any thoughts on this ?